File: RB110/ruby_interview_problems/new_rb119_interview_study_problems/problem17.rb
=begin

Create a method that takes an array of integers as an argument. The method 
should determine the minimum integer value that can be appended to the array 
so the sum of all the elements equal the closest prime number that is greater 
than the current sum of the numbers. For example, the numbers in [1, 2, 3] 
sum to 6. The nearest prime number greater than 6 is 7. Thus, we can add 1 
to the array to sum to 7.

Notes:

The array will always contain at least 2 integers.
All values in the array must be positive (> 0).
There may be multiple occurrences of the various numbers in the array.

p nearest_prime_sum([1, 2, 3]) == 1        # Nearest prime to 6 is 7
p nearest_prime_sum([5, 2]) == 4           # Nearest prime to 7 is 11
p nearest_prime_sum([1, 1, 1]) == 2        # Nearest prime to 3 is 5
p nearest_prime_sum([2, 12, 8, 4, 6]) == 5 # Nearest prime to 32 is 37

# Nearest prime to 163 is 167
p nearest_prime_sum([50, 39, 49, 6, 17, 2) == 4


Problem: Given an array of integers, return an integer

Rules:
- Return integer should be the smallest value that can be
  added to the given array, such that the sum of all nums equal the
  closest prime number that is greater than the curret sum of the numbers

- What is a prime number?
  Any number that is only divisble by 1 and itself
- Closest prime number should be > the sum of the array
- 1 is not a prime number
- If the current num is a prime num, find the next greatest prime num

Algorithm:
- FIND if the current num isa prime num
  - num must be divisible by 1 and itself ONLY
  - Check if the num is evenly divisble by 2 upto itself - 1
  - if the total divisble nums is greater than 1
    - return false
  - else return true
- FIND the sum of the given array
- LOOP over the sum until a prime num is found
  - Increment the sum by +1
    - if its  aprime num, break the loop
    - otherwise, repeat the loop until prime is found
- SUBTRACT the prime num from the original sum
- RETURN the difference

=end

def is_prime?(num)
  factors = []
  2.upto(num) do |factor|
    factors << factor if num % factor == 0
  end

  factors.count == 1
end

def nearest_prime_sum(numbers)
  total = numbers.sum
  
  loop do
    total += 1
    break if is_prime?(total)
  end

  total - numbers.sum
end



p nearest_prime_sum([1, 2, 3]) == 1        # Nearest prime to 6 is 7
p nearest_prime_sum([5, 2]) == 4           # Nearest prime to 7 is 11
p nearest_prime_sum([1, 1, 1]) == 2        # Nearest prime to 3 is 5
p nearest_prime_sum([2, 12, 8, 4, 6]) == 5 # Nearest prime to 32 is 37

# Nearest prime to 163 is 167
p nearest_prime_sum([50, 39, 49, 6, 17, 2]) == 4

File: LS215/Lesson3/problem17.js

/*
You are given an array containing multiple nested arrays of varying depths. 
Each nested array contains numbers, but some may also contain empty subarrays. 
Your goal is to write a function that processes this input and returns a single 
array with all numbers sorted in ascending order.

Do not use the sort method.


const input1 = [[3, [2, 1]], [5, [4, [6]]]];

Output: [1, 2, 3, 4, 5, 6]

const input2 = [[10, 5, [5, 10]], [20, 15], []];

Output: [5, 10, 15, 20]

---- PROBLEM
INPUT: Given with nested subarrays
OUTPUT: A single array with all integers sorted in ascending order, so smallest to largest

Rules:
- Cant use sort() method
- Some subarrays may be empty, others will vcontain various amounts of elements that are ints
- Return a 1 dimensional array

Assumptions:
  - Subarrays will always be number types or empty
  - May include negative numbers, but not other special types of numbers
  - If given an empty array
    - Return an empty array
  - Argument will always be an array
  - Always receive precisely one array arg
  - Remove all duplicate elements, should only contain a single copy

---- EXAMPLES
const input1 = [[3, [2, 1]], [5, [4, [6]]]];
  - Given [[3, [2, 1]], [5, [4, [6]]]]
    - 3, 2, 1, 5, 4, 6
    -> [1, 2, 3, 4, 5, 6]
Output: [1, 2, 3, 4, 5, 6]

const input2 = [[10, 5, [5, 10]], [20, 15], []];

Output: [5, 10, 15, 20]

---- DATA STRUCTURES
- stick w/ array 

---- ALGORITHM
1. Handle any special types of inputs
  - If given empty array
    - Return empty array
2. Create a flattend array of integers
3. SElect all unique integers
3. Sort these ints from least to greatest #cant use sort() ** helper 2
4. Return the sorted array

** helper1
createFlatArray
- Looping over the current array
  - find the length of the array
  - Flatten the array
    - original length of my array and the flatten length of my array are the same
      - Call one final flat on the array
    - if originalLength !== currentLength
      - repeat the loop

** helper2
- Loop over my uniq numbers
  - Init swap to false;
  - Compare the current the element to the next
  - If the current is greater than the next element
    - Flip their position
    - `swap` true
  - If not, 
    - continue onto the next element
- break out if swap set to false
- return the array of sorted numbers
*/

function sortNumbers(numbers) {
  if (numbers.length === 0) return [];

  let flatArray = createFlatArray(numbers);
  let uniqInts = findUniqInts(flatArray);
  return sortInts(uniqInts);
}

function createFlatArray(numbers) {
  let previousLength = numbers.length
  let currentLength = numbers.flat().length;

  while (previousLength !== currentLength) {
    currentLength = numbers.length;
    numbers = numbers.flat();
    previousLength = numbers.length;
  }

  numbers = numbers.flat();
  return numbers;
}

function findUniqInts(numbers) {
  let uniqInts = [];

  numbers.forEach(num => {
    if (!uniqInts.includes(num)) uniqInts.push(num);
  });

  return uniqInts;
}

function sortInts(numbers) {
  let length = numbers.length;
  let swap = true;

  while (swap) {
    swap = false;
    for (let idx = 0; idx < length - 1; idx += 1) {
      let firstNum = numbers[idx];
      let secondNum = numbers[idx + 1];

      if (firstNum > secondNum) {
        numbers[idx] = secondNum;
        numbers[idx + 1] = firstNum;
        swap = true;
      }
    }
    length -= 1;
  }

  return numbers;
}


let input1 = [[3, [2, 1]], [5, [4, [6, [7]]]]]; // Output: [1, 2, 3, 4, 5, 6]
console.log(sortNumbers(input1));
let input2 = [[10, 5, [5, 10]], [20, 15], []]; // Output: [5, 10, 15, 20]
console.log(sortNumbers(input2));

console.log(sortNumbers([]));